module/file.go

```diff
@@ -7,87 +7,89 @@ import (
+}
+
+type ModuleFile struct {
+	Version      uint
+	Layout       string
+	Dependencies map[string]Dependency
+	Flags        map[string]string
+}
+
+// MODULE file version 2
+
+type v2Dependency struct {
+	URL     string
+	Version string
 }
 
-type PinnedDependency struct {
+type v2PinnedDependency struct {
 	URL     string
 	Version string
 	Hash    string
 }
 
-type ModuleFile struct {
+type v2ModuleFile struct {
 	Version            uint
 	Layout             string
-	Dependencies       map[string]Dependency
-	PinnedDependencies map[string]PinnedDependency
+	Dependencies       map[string]v2Dependency
+	PinnedDependencies map[string]v2PinnedDependency
 	Flags              map[string]string
 }
 
-type moduleFileVersion struct {
-	Version uint
-}
+// MODULE file version 1
 
-type legacyVersion struct {
+type v1Version struct {
 	Rev  string
 	Hash string
 }
 
-type legacyDependency struct {
+type v1Dependency struct {
 	Name    string
 	URL     string
-	Version legacyVersion
+	Version v1Version
 }
 
-type legacyModuleFile struct {
-	Dependencies []legacyDependency
+type v1ModuleFile struct {
+	Dependencies []v1Dependency
 }
 
 // ReadModuleFile reads and parses module Dependencies from a MODULE file.
 func ReadModuleFile(modulePath string) ModuleFile {
-	moduleFile := ModuleFile{
-		Version:            util.DbtVersion[1],
-		Dependencies:       map[string]Dependency{},
-		PinnedDependencies: map[string]PinnedDependency{},
-	}
-
 	moduleFilePath := path.Join(modulePath, util.ModuleFileName)
 	if !util.FileExists(moduleFilePath) {
 		log.Debug("Module has no %s file.\n", util.ModuleFileName)
-		return moduleFile
+		return ModuleFile{
+			Version:      util.DbtVersion[1],
+			Dependencies: map[string]Dependency{},
+		}
 	}
 
 	// Check MODULE file version.
 	var moduleFileVersion moduleFileVersion
 	util.ReadYaml(moduleFilePath, &moduleFileVersion)
-	if moduleFileVersion.Version == util.DbtVersion[1] {
-		util.ReadYaml(moduleFilePath, &moduleFile)
-	} else {
-		var legacyModuleFile legacyModuleFile
-		util.ReadYaml(moduleFilePath, &legacyModuleFile)
-		for _, legacyDep := range legacyModuleFile.Dependencies {
-			dep := Dependency{URL: legacyDep.URL, Version: legacyDep.Version.Rev}
-			moduleFile.Dependencies[legacyDep.Name] = dep
-			if legacyDep.Version.Hash != "" {
-				moduleFile.PinnedDependencies[legacyDep.Name] = PinnedDependency{
-					URL:     dep.URL,
-					Version: dep.Version,
-					Hash:    legacyDep.Version.Hash,
-				}
-			}
-		}
-	}
-	// YAML decoding can nil maps if the key is present in the yaml file
-	// with no entries. Restore them if this happens.
-	if moduleFile.Dependencies == nil {
-		moduleFile.Dependencies = map[string]Dependency{}
-	}
-	if moduleFile.PinnedDependencies == nil {
-		moduleFile.PinnedDependencies = map[string]PinnedDependency{}
+
+	switch moduleFileVersion.Version {
+	case 1:
+		return readV1ModuleFile(moduleFilePath)
+	case 2:
+		return readV2ModuleFile(moduleFilePath)
+	case util.DbtVersion[1]:
```

Comment:
So when dbt version becomes 1.4.0, the module files with v3 would fall to default? It seems what you want here instead is `moduleFileVersion.Version <= util.DbtVersion[1]`. Or did I miss something?

Reply:
The idea behind this is that the DBT minor version IS the same as the MODULE file version. So we will only get to v1.4.0 when we introduce a new MODULE file version. This was done to reduce the number of versions DBT has. We already have the version of DBT itself, the MODULE file version, dbt-rules version, and the protocol between DBT and the ninja file generator. But we can also separate these if you think that makes it easier.

Reply:
It just tightly couples the module version to the dbt version, but I'm fine with it. Is that mentioned somewhere in the sources? I think we should have a comment in `util.go` so that people like me who didn't have this context know that updating the dbt minor version should be done when the MODULE version is updated as well.